src/shared/dns-domain.c

```diff
@@ -36,13 +36,15 @@ int dns_label_unescape(const char **name, char *dest, size_t sz) {
         d = dest;
 
         for (;;) {
-                if (*n == '.') {
-                        n++;
-                        break;
-                }
+                if (*n == 0 || *n == '.') {
+                        if (FLAGS_SET(flags, DNS_LABEL_LDH) && r > 0 && n[-1] == '-')
```

Comment:
hmm, it feels a bit sloppy to check the last character in the *source* string, i.e. in n[-1], because we apply unescaping on that after all... (well, we don't allow escapes with your patch, but I think we should). I'd prefer if we'd just have a new local variable `last_char` or so that always contains the last char that was added to the output string, that we can check here.